### PR TITLE
Insert null checks for pointer dereferences when debug assertions are enabled

### PR DESCRIPTION
**Pull request by 1c3t3a**
Similar to how the alignment is already checked, this adds a check
for null pointer dereferences in debug mode. It is implemented similarly
to the alignment check as a `MirPass`.

This inserts checks in the same places as the `CheckAlignment` pass and additionally
also inserts checks for `Borrows`, so code like
```rust
let ptr: *const u32 = std::ptr::null();
let val: &u32 = unsafe { &*ptr };
```
will have a check inserted on dereference. This is done because null references
are UB. The alignment check doesn't cover these places, because in `&(*ptr).field`,
the exact requirement is that the final reference must be aligned. This is something to
consider further enhancements of the alignment check.

For now this is implemented as a separate `MirPass`, to make it easy to disable
this check if necessary.

This is related to a 2025H1 project goal for better UB checks in debug
mode: https://github.com/rust-lang/rust-project-goals/pull/177.

r? @saethlin

### CODE REVIEW

File: compiler/rustc_mir_transform/src/check_pointers.rs [hunk 1]
```diff
@@ -0,0 +1,161 @@
+use rustc_hir::lang_items::LangItem;
+use rustc_index::IndexVec;
+use rustc_middle::mir::visit::{MutatingUseContext, NonMutatingUseContext, PlaceContext, Visitor};
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use tracing::{debug, trace};
+
+pub(crate) fn check_pointers<'a, 'tcx, F>(
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/check_pointers.rs:8]: Please add an extensive doc comment explaining what this does. I currently find this logic too hard to follow.

File: compiler/rustc_mir_transform/src/check_null.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+pub(super) struct CheckNull;
+
+impl<'tcx> crate::MirPass<'tcx> for CheckNull {
+    fn is_enabled(&self, sess: &Session) -> bool {
+        sess.ub_checks()
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        check_pointers(tcx, body, &[], insert_null_check);
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/check_null.rs:17]: Note that zero-sized reads/writes from null pointers are actually allowed. Does your pass handle that correctly?

File: compiler/rustc_mir_transform/src/check_pointers.rs (hunk 1 above)
Comment (saethlin) [reply compiler/rustc_mir_transform/src/check_pointers.rs:8]: FYI, this is a cut-and-paste of the previous implementation. If you think the use of PlaceMention here is confusing, we could indeed revert that. But I am not sure such concerns belong in this PR.

File: compiler/rustc_mir_transform/src/check_null.rs (hunk 2 above)
Comment (saethlin) [reply compiler/rustc_mir_transform/src/check_null.rs:17]: (This is also a missing test case)

File: compiler/rustc_mir_transform/src/check_pointers.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_mir_transform/src/check_pointers.rs:8]: Previously this was an internal implementation detail of one pass. Now it's a shared utility. A shared utility should document its contract with its users.

Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_pointers.rs:28]: Sorry, that's a very fair point. I refactored the interface to be explained better and be more intuitive, PTAL.

File: compiler/rustc_mir_transform/src/check_null.rs (hunk 2 above)
Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_null.rs:17]: Very good point, I added that!

File: compiler/rustc_mir_transform/src/check_null.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        user_ty: None,
+        const_: Const::Val(ConstValue::Scalar(Scalar::from_target_usize(0, &tcx)), tcx.types.usize),
+    }));
+    let is_pointee_no_zst =
+        local_decls.push(LocalDecl::with_source_info(tcx.types.bool, source_info)).into();
+    stmts.push(Statement {
+        source_info,
+        kind: StatementKind::Assign(Box::new((
+            is_pointee_no_zst,
+            Rvalue::BinaryOp(BinOp::Eq, Box::new((Operand::Copy(sizeof_pointee), zero.clone()))),
```
Comment (saethlin) [inline compiler/rustc_mir_transform/src/check_null.rs:62]: The variable says "pointee no ZST" but this is `BinOp::Eq` so I think this is checking if the pointee is a ZST.

File: compiler/rustc_mir_transform/src/check_null.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                BinOp::BitAnd,
+                Box::new((Operand::Copy(is_null), Operand::Copy(is_pointee_no_zst))),
+            ),
+        ))),
+    });
+
+    // Emit a PointerCheck that asserts on the condition and otherwise triggers
+    // a AssertKind::NullPointerDereference.
+    PointerCheck {
+        cond: Operand::Copy(should_throw_exception),
```
Comment (saethlin) [inline compiler/rustc_mir_transform/src/check_null.rs:93]: The condition for `CheckAlignment` is called `is_ok`, not `should_throw`. So one of the pointer-checking transforms is computing the bool wrong. These sorts of bools always turn my brain inside-out, so fixing this by changing this pass or by changing `CheckAlignment` is fine. Whatever convention makes more sense when considering the whole system is good.

**Comment by saethlin** [comment]
For work like this that changes codegen, `x test ui` is almost always better to work off than trying to look at CI, because running the whole UI test suite will exercise your change on a lot of small programs, so if something in your change is broken it's usually pretty easy to narrow it down.

CI always tests using a stage2 build, so if you break codegen, you'll often get a CI failure while using your new compiler to build itself, which is hard to debug from.

File: compiler/rustc_mir_transform/src/check_pointers.rs [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+/// to insert the actual logic for a pointer check (e.g. check for alignment).
+/// This utility takes care of the right order of blocks, the only thing a
+/// caller must do in `on_finding` is:
+/// - Append [Statement]s to `stmts`.
+/// - Append [LocalDecl]s to `local_decls`.
+/// - Return a [PointerCheck] that contains the condition and an [AssertKind].
+///   The AssertKind must be a panic with `#[rustc_nounwind]`.
+/// This utility will insert a terminator block that asserts on the condition
+/// and panics on failure.
+pub(crate) fn check_pointers<'a, 'tcx, F>(
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/check_pointers.rs:28]: I assume by "unsafe pointer" you mean raw pointer? Maybe use standard terminology instead.

File: compiler/rustc_mir_transform/src/check_pointers.rs [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for PointerFinder<'a, 'tcx> {
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, location: Location) {
+        // We want to only check reads and writes to Places, so we specifically exclude
+        // Borrow and RawBorrow.
+        match context {
+            PlaceContext::MutatingUse(
+                MutatingUseContext::Store
+                | MutatingUseContext::AsmOutput
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/check_pointers.rs:132]: Does AsmOutput really require alignment and non-null? (I don't entirely know what it means to be an "asm output operand".)

File: compiler/rustc_mir_transform/src/check_pointers.rs [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+/// fail the assert and an [AssertKind] that defines the behavior on failure.
+pub(crate) struct PointerCheck<'tcx> {
+    pub(crate) cond: Operand<'tcx>,
+    pub(crate) assert_kind: Box<AssertKind<Operand<'tcx>>>,
+}
+
+/// Utility for adding a check for read/write on every sized, unsafe pointer.
+///
+/// Visits every read/write access to a [Sized], unsafe pointer and inserts a
+/// new basic block directly before the pointer access. Then calls `on_finding`
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/check_pointers.rs:18]:
```suggestion
/// new basic block directly before the pointer access. (Read/write accesses are
/// determined by the `PlaceContext` of the MIR visitor. In particular, uses of pointers
/// in borrow expressions are *not* visited.) Then calls `on_finding`
```

**Comment by RalfJung** [comment]
Is there a specific reason that this is a separate MIR pass from the null ptr check? Together they form the pointer validity checks, so I don't quite see why those would be checked separately.

File: compiler/rustc_mir_transform/src/check_null.rs (hunk 4 above)
Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_null.rs:93]: Yes, this blew my head a fair bit, too. I fixed it now and have proper naming and documentation on what it should evaluate to.

File: compiler/rustc_mir_transform/src/check_pointers.rs (hunk 7 above)
Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_pointers.rs:18]: Done

File: compiler/rustc_mir_transform/src/check_pointers.rs (hunk 5 above)
Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_pointers.rs:28]: Yes, makes sense. I guess this originates from `Ty::is_unsafe_ptr`, which basically boils down to a raw pointer. Does it make sense to rename that method as well?

File: compiler/rustc_mir_transform/src/check_pointers.rs (hunk 6 above)
Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_pointers.rs:132]: Judging from the [visitor code](https://github.com/rust-lang/rust/blob/master/compiler/rustc_middle/src/mir/visit.rs#L593-L609) this seems to comes only from inline assembly blocks and their operands (for `out` and `inout` operands). I personally believe this makes sense to also check, but I am happy to be convinced otherwise. Anyways if we believe to keep this we should add a test for it and also for the alignment case.

@saethlin who intially added this `PlaceContext`, would be interesting to get your opinion here!

Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_pointers.rs:132]: Oke I tried to write a test and turns out the current implementation doesn't even sanitize `AsmOutput` places (at least I didn't find code that triggers that). So AFAIC we could leave this away without any effect. 

My hypothesis is that `AsmOutput` places can only occur for `TerminatorKind::InlineAsm` which our current `PointerFinder` doesn't visit since it only looks at a blocks statement.

Maybe we should first decide whether we want to sanitize this or not before investing into implementing it.

**Comment by 1c3t3a** [comment]
> Is there a specific reason that this is a separate MIR pass from the null ptr check? Together they form the pointer validity checks, so I don't quite see why those would be checked separately.

My reasoning here was that this is two separate checks that people maybe want to enable or disable separately (lets say they only like to pay the overhead for alignment and don't care about null). My thinking was that two separate MIR passes solve this problem the most straightforward way, but I am happy to discuss this.

**Comment by 1c3t3a** [comment]
> The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12518233810/job/34920349731) [(plain)](https://github.com/rust-lang/rust/commit/f0ba4d7542616b0d6da9781c927b239bb21b958f/checks/34920349731/logs)
> 
> Click to see the possible cause of the failure (guessed by this bot)

This looks unrelated. Should I just wait and sync again later?

**Comment by RalfJung** [comment]
Your assigned reviewer is @saethlin ; I don't have anything to add here beyond what I already wrote. I still think this should be a single pass doing both null pointer and alignment checks.

**Comment by saethlin** [comment]
> This looks unrelated. Should I just wait and sync again later?

In general, we don't have flaky tests.

This is a clippy test for a lint that is trying to look for functions that do not do any const-incompatible operations. Clippy lints are often implemented in ways that are pretty subtle. So I suspect that this one is analyzing the MIR, and this pass you're working on is causing rustc to generate different different _runtime_ MIR which is no longer const-eligible. If that is the case, the clippy lint is buggy but in a way that's probably fine for a linter (analyzing runtime MIR will produce different results depending on optimization level and other flags), so you can either bless the test to the new behavior as the CI logs say, or see if the clippy test suite has a way to specify flags for the test. If there is such a way, set `-Zub-checks=no` for that test.

File: compiler/rustc_mir_transform/src/check_pointers.rs (hunk 6 above)
Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_pointers.rs]: I chatted with @saethlin and he told me the current unittest-suite is representative of the UB model for the alignment check. Removing `AsmOutput ` doesn't call a test to fail, so I consider it fine to be removed and do that as part of this PR.

**Comment by 1c3t3a** [comment]
> In general, we don't have flaky tests.
This is a clippy test for a lint that is trying to look for functions that do not do any const-incompatible operations.

Thanks for the hint, it actually hits another clippy issue which I try to solve [separately](https://github.com/rust-lang/rust-clippy/pull/14003). Apart from this I did a few changes to the PR:

1. Also insert null-pointer checks for references, as this is also UB.
2. Remove checks for the `AsmOutput` Place, this was done because removing it doesn't break any test for the alignment check, so I assume it is not critical to have this checked. I am happy for feedback around that.
3. Fix various failing tests, althought the clippy test failure is fixed separately and this change needs to be rebased after clippy is fixed.

It would be really nice to get feedback about the following open questions:
1. Should this check be emitted as part of one MIR Pass or is it fine to be two separate ones?
2. Can we drop the `AsmOutput` visit?
3. Is there a performance overhead for also visiting borrows?

**Comment by saethlin** [comment]
> Should this check be emitted as part of one MIR Pass or is it fine to be two separate ones?

It is fine as two separate passes, at least for now. There is a fair chance that in the short term, users want to toggle this new check off and having a separate pass gives an easy compiler flag to do so.

> Can we drop the AsmOutput visit?

Yes. If someone can come up with some UB that we miss by removing this visit, we can figure out a way to add it back. I at least can't think of why this would be required.

> Is there a performance overhead for also visiting borrows?

Almost certainly, but that is a question better answered with empiricism. We have a benchmark suite for compile-time overhead that I or anyone else can kick off upon your request, there is a thread in t-compiler/help called "perf run" that you can post in to ask for a perf run. Just don't force-push while its try-build is running or it'll silently be cancelled.

There is *also* probably a runtime performance overhead, but I suspect that if the compile-time overhead is small the runtime overhead will be too.

**Comment by RalfJung** [comment]
> Is there a performance overhead for also visiting borrows?

Wait, what does this check for borrows? Please extend the PR description to explain that. I expected this to work like the alignment check. Does this change how the alignment check behaves?

File: compiler/rustc_mir_transform/src/check_pointers.rs [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    pub(crate) cond: Operand<'tcx>,
+    pub(crate) assert_kind: Box<AssertKind<Operand<'tcx>>>,
+}
+
+/// Utility for adding a check for read/write on every sized, raw pointer.
+///
+/// Visits every read/write access to a [Sized], raw pointer and inserts a
+/// new basic block directly before the pointer access. (Read/write accesses
+/// are determined by the `PlaceContext` of the MIR visitor. In particular,
+/// uses of pointers in borrow expressions are *not* visited). Then calls
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/check_pointers.rs:20]:
```suggestion
/// uses of pointers in borrow expressions are *not* visited.) Then calls
```

Comment (RalfJung) [inline compiler/rustc_mir_transform/src/check_pointers.rs:20]: This seems to say borrows are indeed not checked so I am confused now.

Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_pointers.rs]: Sorry for the confusion, see answer in the main comment thread.

**Comment by 1c3t3a** [comment]
> Wait, what does this check for borrows?

Sorry for the confusion! You were right that I had an outdated comment in code... I updated it and also updated this PR's description.

> I expected this to work like the alignment check. Does this change how the alignment check behaves?

No it doesn't change any behavior of the alignment check. However, I propose to insert the new alignment check on borrows to prevent code like this to be executed in debug builds:
```rust
let ptr: *const u32 = std::ptr::null();
let val: &u32 = unsafe { &*ptr };
```

File: compiler/rustc_mir_transform/src/check_pointers.rs [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+            PlaceContext::NonMutatingUse(
+                NonMutatingUseContext::Copy | NonMutatingUseContext::Move,
+            ) => {}
+            PlaceContext::MutatingUse(MutatingUseContext::Borrow)
+            | PlaceContext::NonMutatingUse(NonMutatingUseContext::SharedBorrow)
+                if matches!(self.borrow_check_mode, BorrowCheckMode::IncludeBorrows) => {}
+            _ => {
+                return;
+            }
+        }
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/check_pointers.rs:174]: This would be easier to understand if you made it an expression or function returning a `bool` saying whether the place should be visited. Right now it relies on control flow, and combined with the poor formatting that rustfmt enforces here, the empty mach arms (which carry all the weight) are rather hard to read IMO.

**Comment by RalfJung** [comment]
> No it doesn't change any behavior of the alignment check. However, I propose to insert the new null check on borrows to prevent code like this to be executed in debug builds:

What is the justification for adding a null ptr check but no alignment check in that case? Misaligned references are also UB.

> This inserts checks in the same places as the CheckAlignment pass, with
the only difference of also inserting checks for Borrows

So it *doesn't* insert checks in the same places, then. :)

File: tests/ui/mir/null/addrof_null.rs
Comment (RalfJung) [inline tests/ui/mir/null/addrof_null.rs:1]: This is a lot of tests, and they are not fully checking what we actually care about. Specifically, all these tests already pass on rustc master, so they fail to check if the new pass actually does anything.

File: compiler/rustc_mir_transform/src/check_null.rs
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/check_null.rs:1]: If `x` is a reference, does this insert a check for `&*x` or `&x.field`? I think it should not.

File: compiler/rustc_mir_transform/src/check_pointers.rs (hunk 9 above)
Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_pointers.rs:174]: Agree, sounds good!

File: tests/ui/mir/null/addrof_null.rs
Comment (1c3t3a) [reply tests/ui/mir/null/addrof_null.rs:1]: So not all tests pass on rustc master, only the one marked with `run-pass`.

I added these tests since the alignment checks did the same, outlining in what places it expects to **not** fire. For me, adding this new check, it was much clearer what is expected behavior and that there is probably an intent to not check borrows in case of the alignment checks. 

I am happy to remove the tests that shouldn't fail, but I wanna note that I think they were helpful.

**Comment by 1c3t3a** [comment]
> What is the justification for adding a null ptr check but no alignment check in that case? Misaligned references are also UB.

So my initial goal of this PR was to add a new null-check and not correct the alignment check. But I am happy to discuss and pivot here. I tried adding borrows for the alignment check as well, but it failed [tests/ui/mir/alignment/place_computation.rs](https://github.com/rust-lang/rust/blob/6a64e3b89724395d87a394ed9ebd954c474b080a/tests/ui/mir/alignment/place_computation.rs#L4). After checking with @saethlin, he said that is intended because of this paragraph in the reference:

> &/&mut on a place requires the alignment of the field type (or else the program would be “producing an invalid value”), which generally is a less restrictive requirement than being based on an aligned pointer.

So there was no easy way yet to add the check for borrows as well in the alignment case, but happy to investigate more.

File: tests/ui/mir/null/addrof_null.rs
Comment (RalfJung) [reply tests/ui/mir/null/addrof_null.rs:1]: Oh d'oh, I missed that some of them are run-fail. Sorry for that!

Okay, makes sense then. Having a comment in the test explaining the intent would be useful, that can sometimes be hard to figure out 2 years down the road. :)

File: compiler/rustc_mir_transform/src/check_alignment.rs [hunk 10]
```diff
@@ -19,162 +18,45 @@ impl<'tcx> crate::MirPass<'tcx> for CheckAlignment {
...
-            return;
-        }
-
-        // Ensure that this place is based on an aligned pointer.
-        self.pointers.push((pointer, pointee_ty));
-
-        self.super_place(place, context, location);
+        // Skip trivially aligned place types.
+        let excluded_pointees = [tcx.types.bool, tcx.types.i8, tcx.types.u8];
+        check_pointers(
```
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/check_alignment.rs:23]:
```suggestion
        // We have to exclude borrows here: in `&x.field`, the exact
        // requirement is that the final reference must be aligned, but
        // `check_pointers` would check that `x` is aligned, which would be wrong.
        check_pointers(
```

File: compiler/rustc_mir_transform/src/check_null.rs
Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_null.rs:1]: No, this only inserts the check for pointers and not for references.

Comment (RalfJung) [reply compiler/rustc_mir_transform/src/check_null.rs:1]: Do the existing mir-opt tests for the check-align pass cover this? One of them can probably be copied for a corresponding test for check-null.

File: compiler/rustc_mir_transform/src/check_alignment.rs (hunk 10 above)
Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_alignment.rs:23]: Thanks for the wording Ralf!

File: tests/ui/mir/null/addrof_null.rs
Comment (1c3t3a) [reply tests/ui/mir/null/addrof_null.rs:1]: No worries and yes it totally makes sense to comment on this, I added clarification.

File: compiler/rustc_mir_transform/src/check_null.rs
Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_null.rs:1]: The alignment suite doesn't have such a test, however I added one now for the new null check. Happy to extend the alignment suite, once support `Borrow` places there as well!

Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_null.rs:1]: Actually the test I wrote doesn't make sense... I am struggling with coming up with a test-case for that, because I would need a null reference to properly test it (or unaligned for the alignment side), which is instant UB.

Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_null.rs:1]: Ahh nice! I wasn't aware of that, I added one!

**Comment by saethlin** [comment]
@craterbot run mode=build-and-test

**Comment by saethlin** [comment]
Crater says 542 regressions, but crater's log analysis is pretty crude and a lot of ecosystem crates have flaky test suites. So in cases like this where you know what error message you are looking for I advise going to the downloads pages of the report, downloading the tarball that is all of the crates, extracting that, and searching for "null pointer dereference" to find crates that have hit your new check.

At a glance, I see:
```
╰ ➤ rg "null pointer deref" --files-with-matches | cut -d/ -f1 | counts
381 counts
(  1)      300 (78.7%, 78.7%): regressed
(  2)       77 (20.2%, 99.0%): test-fail
(  3)        3 ( 0.8%, 99.7%): build-fail
(  4)        1 ( 0.3%,100.0%): spurious-fixed
```
So only about 79% of the regressions have been correctly categorized as regressions by crater. If you just start looking into crates, some patterns will probably emerge immediately, so this becomes a game of find a pattern and then try to search for failures that don't fit that pattern. If you can. Nobody expects you to look through all 381 crates that have hit this new check.

File: compiler/rustc_mir_transform/src/check_pointers.rs (hunk 5 above)
Comment (Noratrieb) [reply compiler/rustc_mir_transform/src/check_pointers.rs:43]: Yes (not in this PR though)

**Comment by 1c3t3a** [comment]
Thanks a lot for the crater run, this is super interesting data!

I started to go through the crates that failed and quickly noticed that a lot of them have two properties:
1. They already have a warning for dereferencing a null pointer (`#[warn(deref_nullptr)]`).
2. A lot of them have bindings to some C code and thus depend on rust-bindgen.

Checking this with the numbers of failures yields that this is indeed the biggest problem:
Number of crates that have **both** a null pointer error and the warning for dereferencing a null pointer:
```
282 counts                                                                                               
(  1)      235 (83.3%, 83.3%): regressed
(  2)       46 (16.3%, 99.6%): test-fail
(  3)        1 ( 0.4%,100.0%): spurious-fixed
```

This is ~74% of all failures and kinda boils down to: "We told you".

I then started to look at where exactly the issue is happening and I always found the same failing test: `bindgen_test_layout_{struct_name}`! I confirmed this with the count of failures that have the previous condition + a failure in this test:
```
276 counts                                                                                               
(  1)      229 (83.0%, 83.0%): regressed
(  2)       46 (16.7%, 99.6%): test-fail
(  3)        1 ( 0.4%,100.0%): spurious-fixed
```

So what we're seeing in 74% of cases is actually the same error and it lies in generated code. Looking at an example of the generated test I found:

```
#[test]
fn bindgen_test_layout___va_list_tag() {
    ...
    assert_eq!(
        unsafe { &(*(::std::ptr::null::<__va_list_tag>())).gp_offset as *const _ as usize },
        0usize,
        concat!(
            "Offset of field: ",
            stringify!(__va_list_tag),
            "::",
            stringify!(gp_offset)
        )
    );
   ...
}
```

Which actually dereferences a pointer and takes a reference to it! Skimming the bindgen repo I didn't find such code for the definition of the `test_layout` test [1], but I found a PR that removes the wrong code: [2].

This at least explains the higher number of failures. What are the general thoughts regarding this number of failures?

[1]: https://github.com/rust-lang/rust-bindgen/blob/59a43e10b758bd86275aefceae29e874157087d8/bindgen/codegen/mod.rs#L2591-L2701
[2]: https://github.com/rust-lang/rust-bindgen/pull/2203

**Comment by RalfJung** [comment]
Good analysis, thanks!

For the non-bindgen cases, if some of those crates seem like they are still maintained it would be worth filing issues to let them know this is coming.

Cc @emilio @pvdrz for some input from the bindgen side. This would make the bindgen-generated tests fail for old versions of bindgen that do not have https://github.com/rust-lang/rust-bindgen/pull/2203. That's more than 2.5 years ago so hopefully this does not affect any code anyone is still maintaining?

**Comment by saethlin** [comment]
> What are the general thoughts regarding this number of failures?

With all the bindgen tests filtered out, that looks closer to a reasonable number of failures. For reference, we saw 86 regressions from the alignment checks, and ignoring the bindgen situation, I would expect around the same number.

IIRC the bindgen situation comes primarily from people checking in the output of bindgen, which is probably a bad idea, because a sufficiently compatible release from bindgen can't fix them. Even though the root cause is (an old version of) bindgen.

**Comment by 1c3t3a** [comment]
> we saw 86 regressions from the alignment checks, and ignoring the bindgen situation, I would expect around the same number.

Oke I see, yeah if we subtract the bindgen number we are in the same ballbark.

> IIRC the bindgen situation comes primarily from people checking in the output of bindgen, which is probably a bad idea.

I agree that this is a bad idea, but I want to note that for the cases that I looked at (maybe 10 crates), all of them were just depending on bindgen and failed due to that. So there is hope that a simple update could actually fix a lot of customers and they don't need to regenerate their Rust binding code.


Also: Did you reach out to crate owners or the "ecosystem" somehow for the alignment check? I will try to work through the remaining crates and see how many of them still look maintained and eventually open issues, but is there more we can do?

**Comment by saethlin** [comment]
> Did you reach out to crate owners or the "ecosystem" somehow for the alignment check?

I skimmed the lottery

**Comment by saethlin** [comment]
Sigh. GitHub UI shifted around under my finger due to page load. Sorry about that. 

I skimmed the list for heavily-used crates and didn't see any. And in the months after the new check, the largest disruptions were to crates that crater does not test: Firefox and a Windows-specific crate. So in these cases where the list is very long, it would be ideal to have the list sorted by downloads or recent downloads. Filing 80 or so boilerplate issues would be neat, but GitHub explicitly forbids filing issues on other projects using automation, and I don't think doing so manually is the best use of time.

**Comment by 1c3t3a** [comment]
Quick update:

My numbers last time for how many failures happened due to the bindgen-problem were actually wrong: There are **330** crashes due to it, so only **51** crashes for other reasons.

I found another bigger prevalance (14 crashes) in an old version the [`x11_dl`](https://crates.io/crates/x11-dl) crate (over 13 million downloads), which had a [null-backed reference as a definition of a static](https://github.com/AltF02/x11-rs/blob/2c570ee95dec3547bceaf6e4c2686aa31d32605e/x11-dl/src/link.rs#L41-L45). Fortunately the issue was fixed in [a PR](https://github.com/AltF02/x11-rs/pull/165) in 2023.

Apart from that I went ahead and notified a few projects that are bigger about this patch and suggested a change. A common theme I saw was a wrong implementation of the `offset_of!` macro, which creates a null reference as to be seen [here](https://github.com/rustgd/glsl-layout/issues/21).

We will go through this list the next few days and probably open a few more issues/fixes, but I didn't spot any bigger library that is affected so I think we can move forward with this patch from my side.

Are there any remaining points up for discussion before landing this? Are there any plans for rolling this out not as part of the default or as an experiment in any way?

**Comment by RalfJung** [comment]
> Apart from that I went ahead and notified a few projects that are bigger about this patch and suggested a change. A common theme I saw was a wrong implementation of the offset_of! macro, which creates a null reference as to be seen https://github.com/rustgd/glsl-layout/issues/21.

Yeah, wrong `offset_of!` came up a lot in some prior UB detection improvements as well. I;m so happy we finally have an official macro for this that we can point people towards. (And if someone doesn't want to increase the MSRV, https://crates.io/crates/memoffset provides a UB-free implementation for older Rust versions.)

File: compiler/rustc_codegen_ssa/src/mir/block.rs [hunk 11]
```diff
@@ -713,6 +713,11 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                 // and `#[track_caller]` adds an implicit third argument.
                 (LangItem::PanicMisalignedPointerDereference, vec![required, found, location])
             }
+            AssertKind::NullPointerDereference => {
+                // It's `fn panic_null_pointer_dereference()`, and
+                // `#[track_caller]` adds an implicit third argument.
```
Comment (saethlin) [inline compiler/rustc_codegen_ssa/src/mir/block.rs:718]:
```suggestion
                // It's `fn panic_null_pointer_dereference()`,
                // `#[track_caller]` adds an implicit argument.
```

File: src/tools/clippy/clippy_lints/src/missing_const_for_fn.rs [hunk 12]
```diff
@@ -156,7 +156,7 @@ impl<'tcx> LateLintPass<'tcx> for MissingConstForFn {
 
         let mir = cx.tcx.optimized_mir(def_id);
 
-        if let Ok(()) = is_min_const_fn(cx.tcx, mir, &self.msrv)
+        if let Ok(()) = is_min_const_fn(cx.tcx, &mir, &self.msrv)
```
Comment (saethlin) [inline src/tools/clippy/clippy_lints/src/missing_const_for_fn.rs:159]: `mir` should already be a reference here, so I don't think this diff is required, right?

File: src/tools/clippy/src/driver.rs [hunk 13]
```diff
@@ -166,6 +166,8 @@ impl rustc_driver::Callbacks for ClippyCallbacks {
         // MIR passes can be enabled / disabled separately, we should figure out, what passes to
         // use for Clippy.
         config.opts.unstable_opts.mir_opt_level = Some(0);
+        config.opts.unstable_opts.mir_enable_passes =
+            vec![("CheckNull".to_owned(), false)];
```
Comment (saethlin) [inline src/tools/clippy/src/driver.rs:170]: Should we also add CheckAlignment here with exactly the same reasoning?

**Comment by saethlin** [comment]
r=me with comments answered or fixed, and commits squashed down.

You probably need to rebase up because the implementation here now conflicts with https://github.com/rust-lang/rust/pull/128657.

File: src/tools/clippy/clippy_lints/src/missing_const_for_fn.rs (hunk 12 above)
Comment (1c3t3a) [reply src/tools/clippy/clippy_lints/src/missing_const_for_fn.rs:159]: Done, you're right!

File: src/tools/clippy/src/driver.rs (hunk 13 above)
Comment (1c3t3a) [reply src/tools/clippy/src/driver.rs:170]: Fair point! I added it :)

File: compiler/rustc_mir_transform/src/check_pointers.rs (hunk 5 above)
Comment (1c3t3a) [reply compiler/rustc_mir_transform/src/check_pointers.rs:43]: See https://github.com/rust-lang/rust/pull/135994

**Comment by 1c3t3a** [comment]
Thanks you so much! I added all requested changes and split this up in the two commits I had here in the beginning. According to my understanding this PR is now approved, I am not sure if I can request to merge it but let me still try :D 

@bors r=saethlin

**Comment by 1c3t3a** [comment]
Also should this be part of relnotes? I saw the corresponding PR for alignment checks were there and this might break a few people? I am adding the label now, feel free to remove if it doesn't sound like a fit.

@rustbot label: relnotes

**Comment by 1c3t3a** [comment]
Should be fixed!

**Comment by saethlin** [comment]
@rustbot label: +perf-regression-triaged

Slight regression is expected, we're generating more MIR to insert new checks. We have evidence from a crater run that these checks find UB in the wild, which justifies the very small compile time slowdown.
